samza-api/src/main/java/org/apache/samza/operators/windows/Window.java

```diff
@@ -18,32 +18,68 @@
+ * further processing.  Use the {@link Windows} APIs to specify various windowing functions.
  *
- * @param <M>  the type of input {@link MessageEnvelope}
- * @param <WK>  the type of key to the {@link Window}
- * @param <WV>  the type of output value in the {@link WindowOutput}
- * @param <WM>  the type of {@link MessageEnvelope} in the window output stream
+ * <p> A window has the following aspects:
+ * <ul>
+ *   <li> Pane: Every result emitted from a window contains one or more {@link MessageEnvelope}s and is referred
+ *   to as a window pane.
+ *   <li> Key: A {@link Window} can group its results by a key. For instance, A common use-case is to group a
+ *   stream by a specified key over a tumbling time window. In this case, the triggers are evaluated per-key.
+ *   <li> Early and Late Triggers: An early trigger allows to emit early, partial window results speculatively. A late trigger
+ *   allows handling of late data arrivals. Refer to the {@link org.apache.samza.operators.triggers.Triggers} APIs for
+ *   creating early and late triggers.
+ * </ul>
+ *
+ * @param <M> type of input {@link MessageEnvelope}.
+ * @param <K> type of key in the {@link MessageEnvelope} in this {@link org.apache.samza.operators.MessageStream}. If a key is specified,
+ *           results are emitted per-key.
+ * @param <WK> type of key in the {@link Window} output.
+ * @param <WV> type of value stored in the {@link Window}.
+ * @param <WM> type of the {@link Window} result.
  */
-public interface Window<M extends MessageEnvelope, WK, WV, WM extends WindowOutput<WK, WV>> {
+@InterfaceStability.Unstable
+public interface Window<M extends MessageEnvelope, K, WK, WV, WM extends WindowOutput<WK, WV>> {
+
+  /**
+   * Set the early triggers for this {@link Window}.
+   * <p>Use the {@link org.apache.samza.operators.triggers.Triggers} APIs to create instances of {@link Trigger}
+   *
+   * @param trigger the early trigger
+   * @return the {@link Window} function with the early trigger
+   */
+  Window<M, K, WK, WV, WM> setEarlyTrigger(Trigger<M, K, WV> trigger);
 
   /**
-   * Set the triggers for this {@link Window}
+   * Set the late triggers for this {@link Window}.
+   * <p>Use the {@link org.apache.samza.operators.triggers.Triggers} APIs to create instances of {@link Trigger}
    *
-   * @param wndTrigger  trigger conditions set by the programmers
-   * @return  the {@link Window} function w/ the trigger {@code wndTrigger}
+   * @param trigger the late trigger
+   * @return the {@link Window} function with the late trigger
    */
-  Window<M, WK, WV, WM> setTriggers(TriggerBuilder<M, WV> wndTrigger);
+  Window<M, K, WK, WV, WM> setLateTrigger(Trigger<M, K, WV> trigger);
 
   /**
-   * Internal implementation helper to get the functions associated with this Window.
+   * Specifies that results from a pane should be retained after they are emitted. This is applicable when each
```

Comment:
👍  on the suggestion of setAccumulationMode(). More explicit to me.

As for the concept of window panes, a window instance, and window function, we really need to clarify them. To me, the followings make sense:
1) Windows is a factory class to create a Window function
2) Applying a Window function on a stream, we get an instance of window, which defines a group of messages with limited size (keyed or non-keyed), uniquely identifiable by a window key
3) Within an instance of window, each time a trigger is fired, the current set of messages used to generate the output is call a window pane. With multiple firing of triggers in a window, it is possible to have multiple window panes fired w/ the same window key.

I thought the above are what we agreed upon. It is still not clear throughout the documentation here. I would recommend to add the above description to the document for Windows, s.t. users have a clear understanding of those concepts.